File: cli/src/skills.rs
//! Privacy-bounded, opt-in navigation skill cache.
//!
//! Skills deliberately contain no task text, secrets, typed values, screen
//! references, or coordinates. They cache only guarded, non-consequential
//! semantic navigation programs which are freshly grounded before replay.

use crate::error::{AndroidError, Result};
use crate::model::{next_action_id, ActionV1, SnapshotV1};
use crate::session::SessionStore;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;
use std::fs;
use std::path::PathBuf;

const SCHEMA: &str = "tempera.android.skills/v1";
const MAX_SKILLS: usize = 128;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillV1 {
    pub id: String,
    pub task_sha256: String,
    pub start_guard: GuardV1,
    pub program: Vec<SkillActionV1>,
    pub completion: CompletionV1,
    pub learned_at_ms: u128,
    pub successes: u32,
    pub failures: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GuardV1 {
    pub package: String,
    pub activity: String,
    pub contains: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SkillActionV1 {
    pub kind: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub selector: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub key: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub direction: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompletionV1 {
    pub package: String,
    pub exact: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct SkillFileV1 {
    schema_version: String,
    skills: Vec<SkillV1>,
}

#[derive(Debug, Clone)]
pub struct SkillStore {
    path: PathBuf,
}

impl SkillStore {
    pub fn from_environment() -> Result<Self> {
        let store = SessionStore::from_environment()?;
        Ok(Self {
            path: store.root().join("skills.json"),
        })
    }

    pub fn list(&self) -> Result<Vec<SkillV1>> {
        let mut skills = self.load()?;
        skills.sort_by(|left, right| {
            (
                right.successes as i64 - right.failures as i64,
                right.learned_at_ms,
            )
                .cmp(&(
                    left.successes as i64 - left.failures as i64,
                    left.learned_at_ms,
                ))
        });
        Ok(skills)
    }

    pub fn clear(&self, confirmed: bool) -> Result<bool> {
        if !confirmed {
            return Err(AndroidError::InvalidInput(
                "skills clear removes the local cache; rerun with --yes".to_string(),
            ));
        }
        if self.path.is_file() {
            fs::remove_file(&self.path)?;
            Ok(true)
        } else {
            Ok(false)
        }
    }

    pub fn candidates(&self, task: &str, snapshot: &SnapshotV1) -> Result<Vec<SkillV1>> {
        let digest = task_digest(task);
        Ok(self
            .load()?
            .into_iter()
            .filter(|skill| {
                skill.task_sha256 == digest && guard_matches(&skill.start_guard, snapshot)
            })
            .collect())
    }

    pub fn learn(
        &self,
        task: &str,
        start: &SnapshotV1,
        actions: &[ActionV1],
        completion: CompletionV1,
    ) -> Result<Option<SkillV1>> {
        let program = portable_program(actions, start)?;
        if program.is_empty() || completion.exact.is_empty() || completion.package.is_empty() {
            return Ok(None);
        }
        let first_selector = program.iter().find_map(|action| action.selector.clone());
        let guard = GuardV1 {
            package: start.package.clone(),
            activity: start.activity.clone(),
            contains: first_selector.into_iter().collect(),
        };
        if !guard_matches(&guard, start) {
            return Ok(None);
        }
        let digest = task_digest(task);
        let identity = serde_json::to_vec(&(&digest, &guard, &program, &completion))?;
        let id = hex::encode(Sha256::digest(identity))[..24].to_string();
        let mut all = self.load()?;
        if let Some(existing) = all.iter_mut().find(|skill| skill.id == id) {
            existing.successes = existing.successes.saturating_add(1);
            existing.learned_at_ms = SnapshotV1::now_ms();
            let skill = existing.clone();
            self.save(&all)?;
            return Ok(Some(skill));
        }
        let skill = SkillV1 {
            id,
            task_sha256: digest,
            start_guard: guard,
            program,
            completion,
            learned_at_ms: SnapshotV1::now_ms(),
            successes: 1,
            failures: 0,
        };
        all.push(skill.clone());
        if all.len() > MAX_SKILLS {
            all.sort_by_key(|entry| entry.learned_at_ms);
            all.drain(..all.len() - MAX_SKILLS);
        }
        self.save(&all)?;
        Ok(Some(skill))
    }

    pub fn record_success(&self, id: &str) -> Result<()> {
        self.update(id, true)
    }

    pub fn record_failure(&self, id: &str) -> Result<()> {
        self.update(id, false)
    }

    fn update(&self, id: &str, success: bool) -> Result<()> {
        let mut all = self.load()?;
        if let Some(skill) = all.iter_mut().find(|skill| skill.id == id) {
            if success {
                skill.successes = skill.successes.saturating_add(1);
            } else {
                skill.failures = skill.failures.saturating_add(1);
            }
            skill.learned_at_ms = SnapshotV1::now_ms();
            if skill.failures >= skill.successes.saturating_add(2).max(3) {
                all.retain(|candidate| candidate.id != id);
            }
            self.save(&all)?;
        }
        Ok(())
    }

    fn load(&self) -> Result<Vec<SkillV1>> {
        if !self.path.is_file() {
            return Ok(Vec::new());
        }
        let file: SkillFileV1 = serde_json::from_slice(&fs::read(&self.path)?)?;
        if file.schema_version != SCHEMA {
            return Err(AndroidError::InvalidInput(format!(
                "{} has unsupported schema",
                self.path.display()
            )));
        }
        Ok(file.skills.into_iter().filter(valid).collect())
    }

    fn save(&self, skills: &[SkillV1]) -> Result<()> {
        let file = SkillFileV1 {
            schema_version: SCHEMA.to_string(),
            skills: skills.to_vec(),
        };
        let temporary = self.path.with_extension("json.tmp");
        fs::write(&temporary, serde_json::to_vec_pretty(&file)?)?;
        fs::rename(temporary, &self.path)?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            fs::set_permissions(&self.path, fs::Permissions::from_mode(0o600))?;
        }
        Ok(())
    }
}

pub fn replay_action(action: &SkillActionV1, snapshot: &SnapshotV1) -> Result<ActionV1> {
    if let Some(selector) = action.selector.as_deref() {
        if snapshot.node(selector).is_none() {
            return Err(AndroidError::InvalidInput(
                "skill start guard no longer matches a current selector".to_string(),
            ));
        }
    }
    Ok(ActionV1 {
        action_id: next_action_id("skill-action"),
        kind: action.kind.clone(),
        selector: action.selector.clone(),
        text: None,
        key: action.key.clone(),
        direction: action.direction.clone(),
        coordinates: None,
        expected_revision: Some(snapshot.revision),
        expected_state_hash: Some(snapshot.state_hash.clone()),
        metadata: BTreeMap::new(),
    })
}

pub fn completion_matches(completion: &CompletionV1, snapshot: &SnapshotV1) -> bool {
    completion.package == snapshot.package
        && completion.exact.iter().all(|expected| {
            snapshot.nodes.iter().any(|node| {
                node.label.eq_ignore_ascii_case(expected)
                    || node
                        .text
                        .as_deref()
                        .is_some_and(|text| text.eq_ignore_ascii_case(expected))
            })
        })
}

pub fn completion_from_evidence(
    package: &str,
    refs: &[String],
    exact: &[String],
    snapshot: &SnapshotV1,
) -> Option<CompletionV1> {
    let mut labels = exact.to_vec();
    for reference in refs {
        if let Some(node) = snapshot.node(reference) {
            if !node.label.is_empty() {
                labels.push(node.label.clone());
            }
        }
    }
    labels.sort();
    labels.dedup();
    (!package.is_empty() && !labels.is_empty()).then(|| CompletionV1 {
        package: package.to_string(),
        exact: labels,
    })
}

fn portable_program(actions: &[ActionV1], snapshot: &SnapshotV1) -> Result<Vec<SkillActionV1>> {
    if actions.is_empty() || actions.len() > 12 {
        return Ok(Vec::new());
    }
    actions
        .iter()
        .map(|action| {
            if !matches!(
                action.kind.as_str(),
                "tap" | "back" | "home" | "scroll" | "wait"
            ) || action.text.is_some()
                || action.coordinates.is_some()
            {
                return Err(AndroidError::InvalidInput(
                    "not a cacheable navigation action".to_string(),
                ));
            }
            let selector = action
                .selector
                .as_deref()
                .map(|selector| {
                    if selector.starts_with('@') {
                        snapshot
                            .node(selector)
                            .map(|node| node.label.clone())
                            .filter(|label| !label.is_empty())
                            .ok_or_else(|| {
                                AndroidError::InvalidInput(
                                    "cannot cache an unresolved semantic reference".to_string(),
                                )
                            })
                    } else {
                        Ok(selector.to_string())
                    }
                })
                .transpose()?;
            if selector.as_deref().is_some_and(is_sensitive) {
                return Err(AndroidError::InvalidInput(
                    "cannot cache a consequential navigation selector".to_string(),
                ));
            }
            Ok(SkillActionV1 {
                kind: action.kind.clone(),
                selector,
                key: action.key.clone(),
                direction: action.direction.clone(),
            })
        })
        .collect()
}

fn valid(skill: &SkillV1) -> bool {
    !skill.id.is_empty()
        && !skill.task_sha256.is_empty()
        && !skill.program.is_empty()
        && !skill.completion.exact.is_empty()
        && skill.program.iter().all(|action| {
            matches!(
                action.kind.as_str(),
                "tap" | "back" | "home" | "scroll" | "wait"
            ) && action
                .selector
                .as_deref()
                .is_none_or(|selector| !selector.starts_with('@') && !is_sensitive(selector))
        })
}
fn task_digest(task: &str) -> String {
    hex::encode(Sha256::digest(
        task.split_whitespace()
            .collect::<Vec<_>>()
            .join(" ")
            .to_lowercase(),
    ))
}
fn guard_matches(guard: &GuardV1, snapshot: &SnapshotV1) -> bool {
    guard.package == snapshot.package
        && guard.activity == snapshot.activity
        && guard.contains.iter().all(|expected| {
            snapshot
                .nodes
                .iter()
                .any(|node| node.label.eq_ignore_ascii_case(expected))
        })
}
fn is_sensitive(value: &str) -> bool {
    [
        "send",
        "post",
        "publish",
        "buy",
        "purchase",
        "pay",
        "transfer",
        "delete",
        "subscribe",
        "book",
        "order",
        "submit",
    ]
    .iter()
    .any(|word| value.to_ascii_lowercase().contains(word))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::model::{NodeV1, RectV1, CONTROL_SCHEMA_V1};
    fn snapshot(label: &str) -> SnapshotV1 {
        SnapshotV1 {
            schema_version: CONTROL_SCHEMA_V1.to_string(),
            session_id: "s".to_string(),
            serial: "e".to_string(),
            target_kind: "emulator".to_string(),
            package: "demo".to_string(),
            activity: ".Main".to_string(),
            screen: [1, 1],
            revision: 1,
            state_hash: "sha256:x".to_string(),
            captured_at_ms: 0,
            nodes: vec![NodeV1 {
                reference: "@e0".to_string(),
                backend_reference: None,
                role: "Button".to_string(),
                label: label.to_string(),
                text: None,
                content_description: None,
                resource_id: None,
                bounds: RectV1 {
                    left: 0,
                    top: 0,
                    right: 1,
                    bottom: 1,
                },
                enabled: true,
                clickable: true,
                editable: false,
                scrollable: false,
                password: false,
                actions: vec![],
            }],
        }
    }
    #[test]
    fn skills_hash_tasks_and_never_store_task_text() {
        let directory = tempfile::tempdir().unwrap();
        let store = SkillStore {
            path: directory.path().join("skills.json"),
        };
        let secret_task = "Open private customer workspace";
        let action = ActionV1 {
            action_id: "a".to_string(),
            kind: "tap".to_string(),
            selector: Some("@e0".to_string()),
            text: None,
            key: None,
            direction: None,
            coordinates: None,
            expected_revision: Some(1),
            expected_state_hash: Some("sha256:x".to_string()),
            metadata: BTreeMap::new(),
        };
        store
            .learn(
                secret_task,
                &snapshot("Settings"),
                &[action],
                CompletionV1 {
                    package: "demo".to_string(),
                    exact: vec!["Settings".to_string()],
                },
            )
            .unwrap();
        let raw = fs::read_to_string(&store.path).unwrap();
        assert!(!raw.contains(secret_task));
        assert!(raw.contains(&task_digest(secret_task)));
    }
    #[test]
    fn sensitive_and_typed_actions_are_not_cacheable() {
        let typed = ActionV1 {
            action_id: "a".to_string(),
            kind: "type".to_string(),
            selector: Some("@e0".to_string()),
            text: Some("secret".to_string()),
            key: None,
            direction: None,
            coordinates: None,
            expected_revision: None,
            expected_state_hash: None,
            metadata: BTreeMap::new(),
        };
        assert!(portable_program(&[typed], &snapshot("Name")).is_err());
        let tap = ActionV1 {
            action_id: "a".to_string(),
            kind: "tap".to_string(),
            selector: Some("@e0".to_string()),
            text: None,
            key: None,
            direction: None,
            coordinates: None,
            expected_revision: None,
            expected_state_hash: None,
            metadata: BTreeMap::new(),
        };
        assert!(portable_program(&[tap], &snapshot("Buy now")).is_err());
    }
}

File: android_simulator/skills.py
from __future__ import annotations

import copy
import hashlib
import json
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .completion import validate_completion_evidence
from .computer_use import Observation
from .program import guard_matches
from .util import atomic_write


SCHEMA_VERSION = "android-agent-skills.v1"
MAX_SKILLS = 128
_SAFE_ACTIONS = {"tap", "back", "home", "scroll", "launch", "wait"}
_SENSITIVE = {
    "send", "post", "publish", "buy", "purchase", "pay", "transfer", "confirm purchase",
    "delete", "remove account", "factory reset", "subscribe", "book", "order", "submit",
}


def default_skill_path() -> Path:
    return Path.home() / ".android-sim" / "agent-skills-v1.json"


def task_digest(task: str) -> str:
    normalized = " ".join(task.casefold().split()).encode("utf-8")
    return hashlib.sha256(normalized).hexdigest()


def _contains_sensitive(value: str) -> bool:
    normalized = value.casefold()
    return any(term in normalized for term in _SENSITIVE)


def _portable_completion(evidence: Any, observation: Observation) -> dict[str, Any] | None:
    if not isinstance(evidence, dict):
        return None
    value: dict[str, Any] = {}
    if isinstance(evidence.get("package"), str) and evidence["package"]:
        value["package"] = evidence["package"]
    if isinstance(evidence.get("activity"), str) and evidence["activity"]:
        value["activity"] = evidence["activity"]

    exact: list[str] = []
    raw_exact = evidence.get("exact", [])
    if isinstance(raw_exact, str):
        raw_exact = [raw_exact]
    if isinstance(raw_exact, list):
        exact.extend(str(item) for item in raw_exact if isinstance(item, str) and item.strip())

    raw_refs = evidence.get("refs", [])
    if isinstance(raw_refs, str):
        raw_refs = [raw_refs]
    if isinstance(raw_refs, list):
        by_ref = {node.ref: node.label for node in observation.nodes if node.label}
        for ref in raw_refs:
            if isinstance(ref, str) and ref in by_ref:
                exact.append(by_ref[ref])

    exact = list(dict.fromkeys(exact))
    if not exact:
        return None
    value["exact"] = exact
    valid, _, _ = validate_completion_evidence(value, observation)
    return value if valid else None


def _program_is_cacheable(program: Any) -> bool:
    if not isinstance(program, list) or not program or len(program) > 12:
        return False
    for raw_step in program:
        if not isinstance(raw_step, dict) or set(raw_step) - {"when", "action"}:
            return False
        action = raw_step.get("action")
        if not isinstance(action, dict):
            return False
        kind = str(action.get("type", ""))
        if kind not in _SAFE_ACTIONS:
            return False
        if action.get("ref"):
            return False
        if any(key in action for key in ("x", "y", "x1", "y1", "x2", "y2", "text", "secret_ref")):
            return False
        selector = action.get("selector")
        if selector is not None:
            if not isinstance(selector, str) or not selector.strip() or _contains_sensitive(selector):
                return False
        guard = raw_step.get("when")
        if guard is not None and not isinstance(guard, dict):
            return False
    return True


def _start_guard(program: list[dict[str, Any]], observation: Observation) -> dict[str, Any]:
    first = program[0]
    guard = copy.deepcopy(first.get("when") or {})
    action = first.get("action") or {}
    kind = action.get("type")
    selector = action.get("selector")
    if kind in {"tap", "scroll"} and isinstance(selector, str) and selector:
        contains = guard.get("contains", [])
        if isinstance(contains, str):
            contains = [contains]
        if not isinstance(contains, list):
            contains = []
        if selector not in contains:
            contains.append(selector)
        guard["contains"] = contains
        if "package" not in guard and observation.package:
            guard["package"] = observation.package
    return guard


@dataclass(frozen=True)
class Skill:
    id: str
    task_sha256: str
    start_guard: dict[str, Any]
    program: list[dict[str, Any]]
    completion_evidence: dict[str, Any]
    learned_at: int
    successes: int = 1
    failures: int = 0

    def as_json(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "task_sha256": self.task_sha256,
            "start_guard": self.start_guard,
            "program": self.program,
            "completion_evidence": self.completion_evidence,
            "learned_at": self.learned_at,
            "successes": self.successes,
            "failures": self.failures,
        }


class SkillStore:
    def __init__(self, path: Path | None = None):
        self.path = (path or default_skill_path()).expanduser()

    def _load(self) -> list[dict[str, Any]]:
        if not self.path.is_file():
            return []
        try:
            value = json.loads(self.path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            return []
        if not isinstance(value, dict) or value.get("schema_version") != SCHEMA_VERSION:
            return []
        entries = value.get("skills")
        return [item for item in entries if isinstance(item, dict)] if isinstance(entries, list) else []

    def _save(self, entries: list[dict[str, Any]]) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        atomic_write(
            self.path,
            json.dumps({"schema_version": SCHEMA_VERSION, "skills": entries[-MAX_SKILLS:]}, indent=2, sort_keys=True) + "\n",
            mode=0o600,
        )

    def candidates(self, task: str, observation: Observation) -> list[Skill]:
        digest = task_digest(task)
        result: list[Skill] = []
        for raw in self._load():
            if raw.get("task_sha256") != digest:
                continue
            guard = raw.get("start_guard")
            program = raw.get("program")
            evidence = raw.get("completion_evidence")
            if not isinstance(guard, dict) or not _program_is_cacheable(program) or not isinstance(evidence, dict):
                continue
            if not guard_matches(observation, guard):
                continue
            try:
                result.append(Skill(
                    id=str(raw["id"]),
                    task_sha256=digest,
                    start_guard=guard,
                    program=copy.deepcopy(program),
                    completion_evidence=copy.deepcopy(evidence),
                    learned_at=int(raw.get("learned_at", 0)),
                    successes=max(0, int(raw.get("successes", 0))),
                    failures=max(0, int(raw.get("failures", 0))),
                ))
            except (KeyError, TypeError, ValueError):
                continue
        return sorted(result, key=lambda item: (item.successes - item.failures, item.learned_at), reverse=True)

    def learn(
        self,
        *,
        task: str,
        start_observation: Observation,
        program: list[dict[str, Any]],
        final_observation: Observation,
        completion_evidence: Any,
    ) -> Skill | None:
        if not _program_is_cacheable(program):
            return None
        portable = _portable_completion(completion_evidence, final_observation)
        if portable is None:
            return None
        digest = task_digest(task)
        guard = _start_guard(program, start_observation)
        if not guard_matches(start_observation, guard):
            return None
        identity_body = {
            "task_sha256": digest,
            "start_guard": guard,
            "program": program,
            "completion_evidence": portable,
        }
        skill_id = hashlib.sha256(
            json.dumps(identity_body, sort_keys=True, separators=(",", ":")).encode("utf-8")
        ).hexdigest()[:24]
        entries = self._load()
        for item in entries:
            if item.get("id") == skill_id:
                item["successes"] = max(0, int(item.get("successes", 0))) + 1
                item["failures"] = max(0, int(item.get("failures", 0)))
                item["learned_at"] = int(time.time())
                self._save(entries)
                return Skill(
                    skill_id,
                    digest,
                    guard,
                    copy.deepcopy(program),
                    portable,
                    int(item["learned_at"]),
                    int(item["successes"]),
                    int(item["failures"]),
                )
        skill = Skill(
            id=skill_id,
            task_sha256=digest,
            start_guard=guard,
            program=copy.deepcopy(program),
            completion_evidence=portable,
            learned_at=int(time.time()),
        )
        entries.append(skill.as_json())
        self._save(entries)
        return skill

    def record_failure(self, skill_id: str) -> None:
        entries = self._load()
        changed = False
        for item in entries:
            if item.get("id") == skill_id:
                item["failures"] = max(0, int(item.get("failures", 0))) + 1
                changed = True
                # Repeatedly failing skills are removed rather than becoming permanent retry noise.
                if int(item["failures"]) >= max(3, int(item.get("successes", 0)) + 2):
                    item["disabled"] = True
        if changed:
            entries = [item for item in entries if not item.get("disabled")]
            self._save(entries)
